_posts/2023-04-19-making-verification-compelling-visual-verification-feedback-for-dafny.markdown

```diff
@@ -0,0 +1,253 @@
+</div>
+<div id="fig_method_summary"></div>
+[<img src="/blog/assets/images/verification-compelling/hover-c.png" alt="Dafny previous hover messages" style="display:block;margin-left:auto;margin-right:auto;width:500px;max-width:95%;"/>](/blog//blog/assets/images/verification-compelling/hover-c.png)
+<div style="text-align:center;font-style:italic;margin-bottom:2em;">
+Fig 4. Method summary on hover. Note the icon that points to <a href="https://dafny.org/dafny/DafnyRef/DafnyRef#sec-verification-debugging-slow">verification debugging when verification is slow</a> if the number of resource units exceeds a threshold (currently 10M). The costlier assertion batch's first line is given, which is useful if all assertions are verified separatedly using <a href="https://dafny.org/dafny/DafnyRef/DafnyRef#sec-vcs_split_on_every_assert"><code>{:vcs_split_on_every_assert}</code></a>
+</div>
+
+One aspect of program verification is that a single error might cover two or more places in the code. For example, post-conditions are asserted at every exit point of a declaration. Therefore, if this assertion fails, it fails on an exit point, and for the particular post-condition.
+[Figure 3a](#fig_previous_diagnostics) shows that the error was never reported on the post-condition itself before (top image). Following user feedback, on [Figure 3b](#fig_new_diagnostics) I made the error visible on post-conditions, using a red-orange squiggly line in-code &mdash; to indicate it is a secondary error &mdash; and using the regular error icon in the gutter.
+Moreover, because hovering is one of first way users can discover features, I added the following improvements to the hover experience, as shown in the bottom of [Figure 3b](#fig_new_diagnostics):
+
+* We quote the exact code portions that Dafny could not prove
+* On the <b style="color:#3794ff">Error</b> keyword, the usual color for links, I added a link to documentation we also wrote about how to [fix failing assertions](https://dafny.org/latest/DafnyRef/DafnyRef\#sec-verification-debugging).
+* "This is assertion #1 of 2" explains how many assertions there are and where is this assertion roughly located in verification, which proved to be useful for [splitting verification in smaller tasks using annotations](https://dafny.org/latest/DafnyRef/DafnyRef\#sec-assertion-batches).
+* "Resource usage: 9K RU" provides a deterministic metric on how many _resource units_ (RU) the solver used to reach this verification result. Resource units do not fluctuate like time would on multiple iterations, or when running on a different computer, although we found that both are correlated.
+* Hovering a _proved_ assertion reveals similar insights, with a positive message instead of an error, such as "Success: The divisior is never zero".
```

Comment:
```suggestion
* Hovering over a _proved_ assertion reveals similar insights, with a positive message instead of an error, such as "Success: The divisor is never zero".
```